### PR TITLE
feat(badge): refactor badges

### PR DESCRIPTION
**Pull request by alizedebray**

### CODE REVIEW

File: packages/web-demo/src/app/bootstrap/components/badge/badge-demo-page/badge-demo-page.component.ts [hunk 1]
```diff
@@ -1,10 +1,14 @@
...
 @Component({
   selector: 'app-badge-demo-page',
   templateUrl: './badge-demo-page.component.html'
 })
 export class BadgeDemoPageComponent  {
-  codeTemplate = codeTemplate;
-
+  badgeTemplate = badgeTemplate;
+  badgeNestedTemplate = badgeNestedTemplate;
+  badgeInteractiveTemplate = badgeInteractiveTemplate;
```
Comment (gfellerph) [inline packages/web-demo/src/app/bootstrap/components/badge/badge-demo-page/badge-demo-page.component.ts:13]: Can you add the hide functionality for the demo app to make a more interactive example? The functionality does not need to be documented I think, implementation will vary wildly (e.g. if you display a list of applied filters badges will be rendered in a loop anyways).

File: packages/web-demo/src/app/bootstrap/components/badge/badge-interactive-demo/badge-interactive-demo.component.html [hunk 2]
```diff
@@ -0,0 +1,10 @@
+<a href="#" class="badge">Link Badge</a>
+<button type="button" class="badge">Button Badge</button>
+<div class="badge-check">
+  <input class="badge-check-input" type="checkbox" value="" id="CheckableBadge">
+  <label class="badge-check-label" for="CheckableBadge">Checkable Badge</label>
+</div>
```
Comment (gfellerph) [inline packages/web-demo/src/app/bootstrap/components/badge/badge-interactive-demo/badge-interactive-demo.component.html:6]:
```suggestion
</div>
<div class="badge-check">
  <input class="badge-check-input" type="checkbox" value="" id="CheckableBadge">
  <label class="badge-check-label" for="CheckableBadge">
    Checkable Badge
    <span class="badge">10</span>
  </label>
</div>
```
Would be nice to have an example with a nested badge.

File: packages/web-demo/src/app/bootstrap/components/badge/badge-interactive-demo/badge-interactive-demo.component.html [hunk 3]
```diff
@@ -0,0 +1,10 @@
...
+<span class="badge">
+  <span>Dismissible Badge</span>
+  <button type="button" class="btn-close" aria-label="Close"></button>
+</span>
```
Comment (gfellerph) [inline packages/web-demo/src/app/bootstrap/components/badge/badge-interactive-demo/badge-interactive-demo.component.html:10]:
```suggestion
</span>
<span class="badge">
  <span>Dismissible Badge</span>
  <span class="badge">10</span>
  <button type="button" class="btn-close" aria-label="Close"></button>
</span>
```
Same same

File: packages/web-styles/src/components/badge.scss [hunk 4]
```diff
@@ -4,30 +4,132 @@
...
+  &-input:focus-visible + &-label {
+    outline-offset: forms.$input-focus-outline-thickness;
+    outline: forms.$input-focus-outline-thickness solid var(--text-color);
+  }
+
+  &-input:not(:checked) + &-label:hover {
+    @include badge-mx.badge-hover-state;
+  }
+
+  &-input:checked + &-label {
```
Comment (gfellerph) [inline packages/web-styles/src/components/badge.scss:103]: In the designs, the checkbox is hidden. The only way of knowing if the badge is "checked" or no is the color. The visible checkbox adds to the visual complexity but makes it easier to see the status. Do you think it's necessary?

**Review by gfellerph — CHANGES_REQUESTED** [review]

File: packages/web-demo/src/app/bootstrap/components/badge/badge-demo-page/badge-demo-page.component.ts (hunk 1 above)
Comment (oliverschuerch) [reply packages/web-demo/src/app/bootstrap/components/badge/badge-demo-page/badge-demo-page.component.ts:13]: It has not been done for alerts/toast so we don't do it for badges neither.
And we will document this component with storybook anyway in the near future.

**Review by gfellerph — APPROVED** [review]

**Pull request merged**